teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,605 @@
+  warnings, errors, and building on existing Tekton UX (`$(params.<name>.env)` similar to existing `$(workspaces.<name>.path)`)
+- How will security be reviewed and by whom?
+  - TBD
+- How will UX be reviewed and by whom?
+  - TBD
+
+### Drawbacks
+One potential drawbacks to the proposed solution(s) in the TEP is potential friction users will face
+in migrating their Tekton configuration to the new `.env` format. NOTE: migration is only necessary after
+the deprecation period
+
+## Alternatives
+
+<!--
+What other approaches did you consider and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+###  Implicitly Project `Parameters` into the Filesystem of `Steps`
+
+Write `Parameters` in files - `/tekton/parameters/<param-name>` - mounted into `Steps`
+in the same way that `Results` are supplied. The file path for the param
+would be made available via a variable (possibly reusing `$(params.<name>)`.
+
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone $(params.url)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      IFS='' read -d -r PARAM_URL /tekton/params/url
+      git clone ${PARAM_URL}
+```
+
+This approach is also under discussion in the context of changing the way `Result`
+values are stored in [TEP-0086][tep-0086-alt]. In that proposal a sidecar is
+responsible for fetching `Results` from previous `Tasks` and writing them to
+a location on disk for use as `Parameters`.
+
+##### Discussion
+
+This approach aligns well with the proposal in [TEP-0086][tep-0086-prop], and
+brings consistency in how we supply both `Parameters` and `Results` to `Steps`.
+
+###  Combine Elements of Both Explicit Environment Variable and Filesystem Projection Solutions
+
+Building on environment variable alternatives (1.0 & 1.1), and filsystem alternatives (2) Tekton Pipelines could offer an explicit option to the user whether to project as env (language agnostic OR shell format) or file.  There could be different env syntaxes for shell (most used in tasks) vs language agnostic env vars to make the option more UX friendly and backwards compatible for the majority of tasks (which use `sh` or `bash`).
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone $(params.url)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone ${$(params.url1.env)}
+      git clone $(params.url2.env_shell)
+      git clone $(params.url3.path)
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_URL1
+      value: $(params.url1)
+    script: |
+      git clone ${PARAM_URL1}
+      git clone ${PARAM_URL2}
+      IFS='' read -d -r PARAM_URL3 /tekton/params/url3
+      git clone ${PARAM_URL3}
```

Comment:
How would this work for array/object params?

Reply:
I have not investigated as deeply for the `.path` alternative as I have for the `.env` option but I would imagine that the implementation would mirror how this works currently - one file for each used piece of the array.  This was the UX would be more straighforward for users (reference the correct file and you get the desired value).  More investigation would need to be done to see if the downsides (scales poorly w/ large objects/arrays) warrant instead serializing arrays/objects into JSON syntax and then creating helper functions to parse the necessary piece of information instead.  I did not update the document here but if folks prefer this alternative or want more detail here now I can add more information.